### ISSUE TITLE
Routes, resources & urls -> Centralised Services

### ISSUE THREAD

**Issue by ErisDS** [TryGhost/Ghost#9192]
This is going to be a catch-all issue for a whole bunch of refactoring I'm working on. 

Below are 3 specific problems I want to solve, and 3 specific use cases I hope will be achievable at the end.

Note: The goal is to move towards these things being possible internally & in internal apps to see what sort of structures/APIs we end up with. Once the refactoring work is done, we'll address exposing functionality in other ways.

Problems:

- Resource <-> URL 2 way generation E.g. what is the url for a post? which post does a url represent?
- Easy way to generate sitemaps, or other lists of urls, with meta data 
- Having routing be enumerable, and then configurable, extensible, etc

**3 main use cases we currently cannot support**

1. custom home pages
2. rendering a resource at a non-standard URL, e.g. /tag/recipes/ -> /recipes/
3. custom routes and their meta data, e.g. whether they should be in the sitemap, E.g. `/subscribe/`, `/featured/`, `/recipes/`, `/private/`, `/*/amp/`.

The expected outcome of this refactoring work is that we will have new services which control the ability to register routes, describe what happens at that route, and provide meta data about the route. Also that can then manage individual URLs, when they are added or removed from the system, and track some meta data about them. 

Rough task list:

- [x] Apps, register routes
- [x] Centralised concept of what routes are available
- [ ] Centralised concept of resources
- [ ] Centralised concept of controllers
- [ ] Centralised concept of renderers
- [ ] Improve url generation to be driven from shared config
- [ ] Separate out the "url generation" from sitemaps -> url service
- [ ] Ability to manage urls: add/edit/remove + events
- [ ] 2 way generation resource <-> url
- [ ] Decorate routes, urls, controllers etc with meta data they need
- [ ] url utils moved into service
- [ ] App routes: when are they enabled? -> feature service?!

Then:
- [ ] Rewrite sitemap code
- [ ] Revisit controllers, how can we generalise them?
- [ ] Revisit full configuration and API, what else can be improved?
- [ ] Tags & Authors get urls in JSON API responses

Notes:
- ensure we also refactor post.published events to never trigger on import as part of this.
- get rid of the weird secure/setRequestIsSecure everywhere (was #6593)
- figure out how route keywords fit into the route system (was #8492)
- improve what is included in sitemaps (revisit ideas in #4588)

**Comment by ErisDS** [comment]
Closing in favour of #9601
